BAMHeader not set when running on a cluster

When running `transform` with `.bam` file as the output, I see the following error:
```
Exception in thread "main" java.lang.AssertionError: assertion failed: Cannot return header if not attached.
        at scala.Predef$.assert(Predef.scala:179)
        at org.bdgenomics.adam.rdd.read.ADAMBAMOutputFormat$.getHeader(ADAMBAMOutputFormat.scala:60)
        at org.bdgenomics.adam.rdd.read.ADAMBAMOutputFormat.<init>(ADAMBAMOutputFormat.scala:68)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at java.lang.Class.newInstance(Class.java:379)
        at org.apache.spark.rdd.InstrumentedOutputFormat.<init>(InstrumentedOutputFormat.scala:33)
        at org.bdgenomics.adam.rdd.read.InstrumentedADAMBAMOutputFormat.<init>(ADAMBAMOutputFormat.scala:71)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at java.lang.Class.newInstance(Class.java:379)
        at org.apache.spark.rdd.PairRDDFunctions.saveAsNewAPIHadoopDataset(PairRDDFunctions.scala:958)
```

I tried to fix this with a similar strategy used by @fnothaft for the VCF header (see https://github.com/bigdatagenomics/adam/compare/master...hammerlab:bam-save-header?expand=1), but did not seem to take.  Any ideas on this?

heh, seems like the errors there were from double- and triple-setting the header on each executor (once per partition), e.g.:
```
java.lang.AssertionError: assertion failed: Cannot attach a new SAM header without first clearing the header.
        at scala.Predef$.assert(Predef.scala:179)
        at org.bdgenomics.adam.rdd.read.ADAMSAMOutputFormat$.addHeader(ADAMSAMOutputFormat.scala:41)
        at org.bdgenomics.adam.rdd.read.AlignmentRecordRDDFunctions$$anonfun$adamSAMSave$1$$anonfun$2.apply(AlignmentRecordRDDFunctions.scala:144)
        at org.bdgenomics.adam.rdd.read.AlignmentRecordRDDFunctions$$anonfun$adamSAMSave$1$$anonfun$2.apply(AlignmentRecordRDDFunctions.scala:134)
```

Forcibly clearing and then re-setting the header on each partition seems to do the trick, e.g.
```
    val mp = rdd.mapPartitionsWithIndex((idx, iter) => {
      log.warn(s"Setting ${if (asSam) "SAM" else "BAM"} header for partition $idx")
      val header = bcastHeader.value
      synchronized {
        // perform map partition call to ensure that the VCF header is set on all
        // nodes in the cluster; see:
        // https://github.com/bigdatagenomics/adam/issues/353

        asSam match {
          case true  =>
            ADAMSAMOutputFormat.clearHeader()
            ADAMSAMOutputFormat.addHeader(header)
            log.warn(s"Set SAM header for partition $idx")
          case false =>
            ADAMBAMOutputFormat.clearHeader()
            ADAMBAMOutputFormat.addHeader(header)
            log.warn(s"Set BAM header for partition $idx")
        }
      }
      Iterator[Int]()
    }).count()
```

I'll open up a PR promptly.

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/features/FeatureParser.scala
@@ -153,8 +153,12 @@ class IntervalListParser extends Serializable {
             val a = fields(4).split(Array(';', ',')).map(field => field.split('|') match {
               case Array(key, value) =>
                 key match {
-                  case "gn" | "ens" | "vega" | "ccds" => (Some(Dbxref.newBuilder().setDb(key).setAccession(value).build()), None)
-                  case _                              => (None, Some(key -> value))
+                  case "gn" | "ens" | "vega" | "ccds" =>
+                    (
+                      Some(Dbxref.newBuilder().setDb(key).setAccession(value).build()),
+                      None
+                    )
+                  case _ => (None, Some(key -> value))
                 }
               case x => throw new Exception(s"Expected fields of the form 'key|value;' but got: $field. Line:\n$line")
             })

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rdd/read/AlignmentRecordRDDFunctions.scala
@@ -130,10 +130,42 @@ class AlignmentRecordRDDFunctions(rdd: RDD[AlignmentRecord])
     // add keys to our records
     val withKey = convertRecords.keyBy(v => new LongWritable(v.get.getAlignmentStart))
 
+    val bcastHeader = rdd.context.broadcast(header)
+    val mp = rdd.mapPartitionsWithIndex((idx, iter) => {
+      log.warn(s"Setting ${if (asSam) "SAM" else "BAM"} header for partition $idx")
+      val header = bcastHeader.value
+      synchronized {
+        // perform map partition call to ensure that the VCF header is set on all
+        // nodes in the cluster; see:
+        // https://github.com/bigdatagenomics/adam/issues/353
+
+        asSam match {
+          case true =>
+            ADAMSAMOutputFormat.clearHeader()
+            ADAMSAMOutputFormat.addHeader(header)
+            log.warn(s"Set SAM header for partition $idx")
+          case false =>
+            ADAMBAMOutputFormat.clearHeader()
+            ADAMBAMOutputFormat.addHeader(header)
+            log.warn(s"Set BAM header for partition $idx")
+        }
+      }
+      Iterator[Int]()
+    }).count()
+
+    // force value check, ensure that computation happens
+    if (mp != 0) {
+      log.warn("Had more than 0 elements after map partitions call to set VCF header across cluster.")
+    }
+
     // attach header to output format
     asSam match {
-      case true  => ADAMSAMOutputFormat.addHeader(header)
-      case false => ADAMBAMOutputFormat.addHeader(header)
+      case true =>
+        ADAMSAMOutputFormat.clearHeader()
+        ADAMSAMOutputFormat.addHeader(header)
+      case false =>
+        ADAMBAMOutputFormat.clearHeader()
+        ADAMBAMOutputFormat.addHeader(header)
     }
 
     // write file to disk

--- a/adam-core/src/main/scala/org/bdgenomics/adam/rich/RichAlignmentRecord.scala
+++ b/adam-core/src/main/scala/org/bdgenomics/adam/rich/RichAlignmentRecord.scala
@@ -68,7 +68,7 @@ class IlluminaOptics(val tile: Long, val x: Long, val y: Long) {}
 
 class RichAlignmentRecord(val record: AlignmentRecord) {
 
-  lazy val referenceLength: Int = RichAlignmentRecord.referenceLengthFromCigar(record.getCigar.toString)
+  lazy val referenceLength: Int = RichAlignmentRecord.referenceLengthFromCigar(record.getCigar)
 
   // Returns the quality scores as a list of bytes
   lazy val qualityScores: Array[Int] = {
@@ -76,7 +76,7 @@ class RichAlignmentRecord(val record: AlignmentRecord) {
   }
 
   // Parse the tags ("key:type:value" triples)
-  lazy val tags: Seq[Attribute] = AttributeUtils.parseAttributes(record.getAttributes.toString)
+  lazy val tags: Seq[Attribute] = AttributeUtils.parseAttributes(record.getAttributes)
 
   // Parses the readname to Illumina optics information
   lazy val illuminaOptics: Option[IlluminaOptics] = {

